### ISSUE TITLE
legion: dependent partitioning operation appears to non-deterministically fail on multiple nodes

### ISSUE THREAD

**Issue by rohany** [StanfordLegion/legion#1172]
I have an application that at it's core looks like this:

```
A = region(Rect<1>, {0, dim})
B = region(int32_t, {0, 1 << 63})
for (i in range(0, numIter):
  do some computation to fill A with ranges into B
  A_part = equal partition of A
  B_part = partition_by_image_range(A_part)
  do some comptutation with A_part and B_part
```

I've run this computation on 1 node with 8 partitions and don't see any errors. I tried scaling this code to 4 nodes (where 8 partitions are used), and non-deterministically see an error. The error is that sometimes, some of the B_part subregions are empty, when they should not be (i.e. the result of partitioning appears to change non-deterministically on multiple nodes). I've run the program with debug on, bounds checks on, -lg:partcheck and -lg:safe_ctrlrepl and seen the error. I've verified that on 4 nodes, the source partitions of A that are resulting in empty partitions of B do indeed contain non-empty rectangles, so B_part should never contain empty subregions.

I'm working on creating a reproducer on sapling for this bug, but unfortunately I've only seen it on some larger test cases running on Lassen. In the meantime, are there logging statements or debug flags that might help tracking this down? 

Edit: I'm having difficulty reproducing the issue on sapling, mainly as all the computation is happening much slower than on Lassen.

This code is running with control replication.

cc-ing @lightsighter for ideas.

**Comment by streichler** [comment]
@rohany how big is `dim` on the runs that fail?
Also, is that definition for `B` right?  `1<<63` probably doesn't mean what you want it to for `int64_t`, let alone `int32_t`.

Logging may be able to shed some light, but it'll need a bit of help to be useful.  Please:
1) add some application level logging to record the ranges you write into A (both the index in A and the range in B, and do it right where you store the range into A)
2) add a logging message when a subregion of B turns out to be empty (please say which subregion)
3) run with that logging enabled and also with `-level part=2,dpops=2`
4) put the logs on sapling and I'll take a look

**Comment by rohany** [comment]
Some clarification before I head off an do this:

> Also, is that definition for B right? 1<<63 probably doesn't mean what you want it to for int64_t, let alone int32_t.

It's not actually 1 << 63 (trying to put it too close to coord_t max raises some other problems in the runtime...). It's just some value larger than I will actually ever need.

> add some application level logging to record the ranges you write into A (both the index in A and the range in B, and do it right where you store the range into A)

The A region is relatively large (~22 million elements). I don't know if this will be tractable to look at. Do you still want it?

> add a logging message when a subregion of B turns out to be empty (please say which subregion)

What do you mean by "please say which subregion"? Is there a particular format of identifier you want?

**Comment by streichler** [comment]
>The A region is relatively large (~22 million elements). I don't know if this will be tractable to look at. Do you still want it?

It'll be a big log file, but without that information, I won't know if the ranges appearing in the deppart log messages are right or not.  The alternative would be to see if you can get the same failure if you use some sort of easily-derived-from-the-index-in-A ranges.  Let's try the logging first though.

> What do you mean by "please say which subregion"? Is there a particular format of identifier you want?

I just need to know which of the subregions of B are empty - the log messages will give me the realm IDs of the subregions in ascending color order, and I need to know which one(s) are wrong.

**Comment by rohany** [comment]
Closing because I found the problem -- it turned out not to have anything to do with dependent partitioning at all. Instead, my compiler had incorrectly generated an inline mapping of a region with RW privilege, and then launched a task on that region also with an RW privilege. It then seemed like legion got confused about which instance to give to the dependent partitioning operation, resulting in a sometimes incorrect (for my application) resulting partition. 

I originally didn't suspect something like this because legion has usually warned me / error out when inline mappings conflict like this. If this sounds like a bug, I can repurpose this issue with a reproducer.

**Comment by manopapad** [comment]
AFAIK if you try to launch a task on a region that's inline mapped, then Legion will automatically unmap the inlined region before the task starts, and remap it after the task finishes. It sounds like this wasn't quite the case for you (although the intervening operation was a partitioning operation, not a task launch). It would be interesting to compare spy graphs of a successful and an unsuccessful run, and possibly also the inputs & output to any `map_partition` calls to your mapper.

**Comment by rohany** [comment]
Actually, there was a task launch in the middle. The full flow of operations was:

create region A
(accidentally) inline map A with RW
launch task over A with RW
use A to fill information in region B <- this step was sometimes getting incorrect data for A (in particular uninitialized data)
compute partitions of region C using image_range on B

**Comment by manopapad** [comment]
In that case you can skip region B, region C and the partitioning operation from your minimal example. Just print out the contents of A right after the task launch, and get spy graphs from a run where A was properly initialized, and one where it was not.

**Comment by rohany** [comment]
Do the spy graphs need legion to be built with -DLEGION_SPY or without is fine?

**Comment by manopapad** [comment]
It should be fine to just run with `-lg:spy -logfile spy_%.log`, no need to recompile with `-DLEGION_SPY`.

**Comment by rohany** [comment]
Lame, both runs failed with:
```
Traceback (most recent call last):
  File "../legion/legion/tools/legion_spy.py", line 13613, in <module>
    main(temp_dir)
  File "../legion/legion/tools/legion_spy.py", line 13494, in main
    state.post_parse(simplify_graphs, physical_checks or event_graphs)
  File "../legion/legion/tools/legion_spy.py", line 12202, in post_parse
    replicant.update_shards()
  File "../legion/legion/tools/legion_spy.py", line 9078, in update_shards
    shard.merge(self.orig)
  File "../legion/legion/tools/legion_spy.py", line 8316, in merge
    assert not other.operations
AssertionError
```

Attaching all of the files here:

Logs from an invalid run:
[spadd3-invalid-spy-0.log](https://github.com/StanfordLegion/legion/files/7868399/spadd3-invalid-spy-0.log)
[spadd3-invalid-spy-1.log](https://github.com/StanfordLegion/legion/files/7868401/spadd3-invalid-spy-1.log)
[spadd3-invalid-spy-2.log](https://github.com/StanfordLegion/legion/files/7868402/spadd3-invalid-spy-2.log)
[spadd3-invalid-spy-3.log](https://github.com/StanfordLegion/legion/files/7868403/spadd3-invalid-spy-3.log)

Logs from a valid run:

[spadd3-valid-spy-0.log](https://github.com/StanfordLegion/legion/files/7868404/spadd3-valid-spy-0.log)
[spadd3-valid-spy-1.log](https://github.com/StanfordLegion/legion/files/7868405/spadd3-valid-spy-1.log)
[spadd3-valid-spy-2.log](https://github.com/StanfordLegion/legion/files/7868406/spadd3-valid-spy-2.log)
[spadd3-valid-spy-3.log](https://github.com/StanfordLegion/legion/files/7868407/spadd3-valid-spy-3.log)

**Comment by manopapad** [comment]
^ @lightsighter these files made Legion spy sad

**Comment by lightsighter** [comment]
Did the run exit cleanly without crashing?

**Comment by lightsighter** [comment]
> I originally didn't suspect something like this because legion has usually warned me / error out when inline mappings conflict like this. If this sounds like a bug, I can repurpose this issue with a reproducer.

Legion should unmap and remap for you. Please make a reproducer.

**Comment by rohany** [comment]
> Did the run exit cleanly without crashing?

Yes

> (assuming you didn't disable its checks by telling it to assume safe launches)

I don't think I did anything like this, but what would indicate I did?

> Please make a reproducer.

I'm not having much luck making a reproducer on sapling. Are there instead log files I can give you from runs on lassen?

**Comment by lightsighter** [comment]
@rohany Pull and regenerate the logs. This time with the runtime built to dump detailed Legion Spy logging. These logs currently give this output when we attempt to run verification on them:

```
WARNING: Requested logical analysis but logging information is missing. Please compile the runtime with USE_SPY=1 to enable validation of the runtime. Disabling logical checks.
WARNING: Requested physical analysis but logging information is missing. Please compile the runtime with USE_SPY=1 to enable validation of the runtime. Disabling physical checks.
```

**Comment by rohany** [comment]
New log files are attached. Running spy on both doesn't seem to give an error, but it complains that the runs may have crashed (which they didn't). I changed the application to just print when it sees the signs of the invalid source regions of the partition instead of crashing.

[spadd3-spy-invalid.zip](https://github.com/StanfordLegion/legion/files/7874450/spadd3-spy-invalid.zip)
[spadd3-spy-valid.zip](https://github.com/StanfordLegion/legion/files/7874451/spadd3-spy-valid.zip)

**Comment by rohany** [comment]
Ah, forgot to run the legion spy with `-lp`. With that, Legion spy reports an error that sounds like exactly what happened:

```
ERROR: No dataflow path found to update field Field 2 of instance Instance 0x4000400040800001 of region requirement 0 of Mapping Op 33
```

**Comment by lightsighter** [comment]
I'm pretty sure the issue is not what you think it is. This is just a duplicate of #907 with inline mappings instead of attach operations. The problem is the same. Singular attach operations and inline mappings in control replicated contexts are just broken in the same way which is clearly noted in #765 and will be until the collective instance branch from #546 is merged.

**Comment by lightsighter** [comment]
We'll keep this issue open for now though to figure out why Legion Spy thinks things were not completely logged.

**Comment by rohany** [comment]
From scanning those linked issues again, I'm not sure I understand such operations are even allowed to happen in control replicated contexts if they are broken -- an error from the runtime instead of this non-deterministic output in a region would have been significantly easier to debug.

**Comment by lightsighter** [comment]
@rohany Are we done with this issue now that collective views have merged (#546)?

**Comment by rohany** [comment]
This is an old way of doing file IO I had in DISTAL but i don't do it this way anymore. I can re-open the issue if I run into anything similar again.
